keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -0,0 +1,1543 @@
+instead fail. By favoring type safety here were are trading off avaiability for
+consistent and predictable behavior.
+
+Example: Typesafe access to object 
+
+```yaml
+apiVersion: admissionregistration.k8s.io/v1
+kind: ValidatingAdmissionPolicy
+metadata:
+  name: "validate-xyz.example.com"
+spec:
+  constraints:
+   rules:
+    - apiGroups:   ["apps"]
+      apiVersions: ["v1"]
+      operations:  ["CREATE", "UPDATE"]
+      resources:   ["deployments"]
+      scope:       "*"
+  validations:
+    # replicas is accessible because this resource matches only v1 deployments
+  - rule: "object.spec.replicas < 100"
+```
+
+Example: Typesafe access only to metadata
+
+```yaml
+apiVersion: admissionregistration.k8s.io/v1
+kind: ValidatingAdmissionPolicy
+metadata:
+  name: "validate-xyz.example.com"
+spec:
+  constraints:
+   rules:
+    - apiGroups:   ["*"]
+      apiVersions: ["*"]
+      operations:  ["CREATE", "UPDATE"]
+      resources:   ["*"]
+      scope:       "*"
+  validations:
+    # minReadySeconds is not accessible because this resource matches multiple types
+  - rule: "object.spec.minReadySeconds > 60" # ERROR! Not such field "minReadySeconds".
+    # metadata is always accessible
+  - rule: "object.name.startsWith('xyz')"
+```
+
+Alternative Considered: Do not use CEL type-checking and instead dynamically evaluate
+all CEL expressions.
+
+Pros:
+
+- Can write a single CEL expression that processes multiple GVKs
+
+Cons:
+
+- All errors are reported during admission control validation. If instead
+  type-checking is used, CEL type-check errors can be surfaced much earlier in
+  the status of `ValidatingAdmissionPolicy` resources when the
+  `ValidatingAdmissionPolicy`, configuration CRDs and configuration resources
+  are written to the API Server. This provides a workflow where type-checking
+  can be used to detect errors early, preventing them from occurring during
+  admission control validation and making it safer to operate a policy using
+  `failureMode: Fail`.
+
+#### Failure Policy
+
+For in-process validation there is no remote request, so errors should be
+deterministic. We believe this significantly reduces the risk of "fail closed"
+admission control as compared to webhooks.
+
+Because failure policy is most often selected based on the need to guarantee
+enforcement, we propose defaulting failure policy to "fail" and allowing it to
+be configured on a per-rule basis:
+
+```yaml
+  validations:
+    - rule: "object.spec.xyz == configuration.x"
+      failurePolicy: Ignore # The default is "Fail"
+```
+
+TODO: Metric for ignored failures?
+
+#### Safety measures
+
+To prevent clusters from being put into a unustable state that cannot be
+recoverd from via the API, admission webhooks are not allowed to match
+`ValidatingWebhookConfiguration` and `MutatingWebhookConfiguration` kinds.
+
+We will extend this approach, and not allow `ValidatingAdmissionPolicy`,
```

Comment:
I think that this is problematic. It makes sense for `ValidatingAdmissionPolicy` to not be applied to itself, but there's no reason why they can't apply to `ValidatingWebhookConfiguration` (or vice versa)

e.g. `ValidatingAdmissionPolicy` applies to everything except `ValidatingAdmissionPolicy`, `ValidatingWebhookConfiguration` applies to everything except `ValidatingWebhookConfiguration` _and_ `ValidatingAdmissionPolicy`

This closes a gap today where `ValidatingWebhookConfigurations` can not be restricted to targetting specific hosts.

Reply:
I agree. I've updated the KEP to match. I suppose this approach could be generalized in the future if needed. Maybe each `ValidatingAdmissionPolicy` can specify a level and a `ValidatingAdmissionPolicy` of a lower level could match a `ValidatingAdmissionPolicy` of a higher level, eliminating the possibility of a cycle like done here between webhooks and cel.